### ISSUE TITLE
Compilation error with Clang and MKL (warnings with GCC)

### ISSUE THREAD

**Issue by avaucher** [davisking/dlib#746]
The simple program `main.cpp`

```c++
#include <mkl.h>
#include <dlib/matrix.h>

int main() {
  return 0;
}
```

fails compiling with Clang 3.9.1:

`clang++ -std=c++11 -I/path/to/mkl/include -I/path/to/dlib/include main.cpp
`
Error:
```
In file included from main.cpp:2:
In file included from /path/to/dlib/include/dlib/matrix/../optimization/../matrix.h:11:
In file included from /path/to/dlib/include/dlib/matrix/matrix_la.h:13:
In file included from /path/to/dlib/include/dlib/matrix/matrix_lu.h:11:
/path/to/dlib/include/dlib/matrix/matrix_trsm.h:15:18: error: conflicting types for 'cblas_strsm'
            void cblas_strsm(const enum CBLAS_ORDER Order, const enum CBLAS_SIDE Side,
                 ^
/path/to/mkl/include/mkl_cblas.h:560:6: note: previous declaration is here
void cblas_strsm(const  CBLAS_LAYOUT Layout, const  CBLAS_SIDE Side,
     ^
In file included from main.cpp:2:
In file included from /path/to/dlib/include/dlib/matrix/../optimization/../matrix.h:11:
In file included from /path/to/dlib/include/dlib/matrix/matrix_la.h:13:
In file included from /path/to/dlib/include/dlib/matrix/matrix_lu.h:11:
/path/to/dlib/include/dlib/matrix/matrix_trsm.h:21:18: error: conflicting types for 'cblas_dtrsm'
            void cblas_dtrsm(const enum CBLAS_ORDER Order, const enum CBLAS_SIDE Side,
                 ^
/path/to/mkl/include/mkl_cblas.h:602:6: note: previous declaration is here
void cblas_dtrsm(const  CBLAS_LAYOUT Layout, const  CBLAS_SIDE Side,
     ^
2 errors generated.

```

---

With GCC 6.2.0, there are warnings but no errors:
`c++ -std=c++11 -I/path/to/mkl/include -I/path/to/dlib/include main.cpp
`

```
In file included from /path/to/dlib/include/dlib/matrix/matrix_lu.h:11:0,
                 from /path/to/dlib/include/dlib/matrix/matrix_la.h:13,
                 from /path/to/dlib/include/dlib/matrix.h:11,
                 from main.cpp:2:
/path/to/dlib/include/dlib/matrix/matrix_trsm.h:19:53: warning: declaration of 'void dlib::blas_bindings::cblas_strsm(dlib::blas_bindings::CBLAS_ORDER, dlib::blas_bindings::CBLAS_SIDE, dlib::blas_bindings::CBLAS_UPLO, dlib::blas_bindings::CBLAS_TRANSPOSE, dlib::blas_bindings::CBLAS_DIAG, int, int, float, const float*, int, float*, int)' with C language linkage
                              float *B, const int ldb);
                                                     ^
In file included from /path/to/mkl/include/mkl.h:38:0,
                 from main.cpp:1:
/path/to/mkl/include/mkl_cblas.h:560:6: warning: conflicts with previous declaration 'void cblas_strsm(CBLAS_LAYOUT, CBLAS_SIDE, CBLAS_UPLO, CBLAS_TRANSPOSE, CBLAS_DIAG, int, int, float, const float*, int, float*, int)'
 void cblas_strsm(const  CBLAS_LAYOUT Layout, const  CBLAS_SIDE Side,
      ^~~~~~~~~~~
In file included from /path/to/dlib/include/dlib/matrix/matrix_lu.h:11:0,
                 from /path/to/dlib/include/dlib/matrix/matrix_la.h:13,
                 from /path/to/dlib/include/dlib/matrix.h:11,
                 from main.cpp:2:
/path/to/dlib/include/dlib/matrix/matrix_trsm.h:25:54: warning: declaration of 'void dlib::blas_bindings::cblas_dtrsm(dlib::blas_bindings::CBLAS_ORDER, dlib::blas_bindings::CBLAS_SIDE, dlib::blas_bindings::CBLAS_UPLO, dlib::blas_bindings::CBLAS_TRANSPOSE, dlib::blas_bindings::CBLAS_DIAG, int, int, double, const double*, int, double*, int)' with C language linkage
                              double *B, const int ldb);
                                                      ^
In file included from /path/to/mkl/include/mkl.h:38:0,
                 from main.cpp:1:
/path/to/mkl/include/mkl_cblas.h:602:6: warning: conflicts with previous declaration 'void cblas_dtrsm(CBLAS_LAYOUT, CBLAS_SIDE, CBLAS_UPLO, CBLAS_TRANSPOSE, CBLAS_DIAG, int, int, double, const double*, int, double*, int)'
 void cblas_dtrsm(const  CBLAS_LAYOUT Layout, const  CBLAS_SIDE Side,
      ^~~~~~~~~~~

```

**Comment by avaucher** [comment]
It relates to https://stackoverflow.com/questions/30939724/error-linking-clang-with-dlib-and-intel-mkl but I found no solution there that does not involve changing the dlib headers.

**Comment by davisking** [comment]
You either don't include the mkl file before dlib, or add another #ifdef
check like this one
https://github.com/davisking/dlib/blob/master/dlib/matrix/matrix_trsm.h#L13
to dlib's header files to avoid declaring this when the mkl file is
included.  Other than that I don't know another reasonable solution.  If
you do the second you should submit a PR with it :)

**Comment by avaucher** [comment]
Currently the error is identical if I switch the order of the include statements.

What I wouldn't like about an additional #ifdef is that then the order of the include statements matters...

@davisking What is the use of `extern "c"` here? If I delete it it compiles without problem.

**Comment by davisking** [comment]
You can't do that, extern "c" is an important statement.  You can lookup what it does in any C++ language resource.

**Comment by avaucher** [comment]
Maybe I should reformulate:  
why have an `extern "c"` here, since dlib is a C++ library? Is there a use case where projects in another programming language are linked against dlib and functions like `cblas_dtrsm` are needed?

**Comment by davisking** [comment]
cblas_dtrsm is a BLAS function, it's not in dlib.  Importantly, cblas_dtrsm
is a C function with C linkability, so extern "c" is required around the
function prototype.

**Comment by avaucher** [comment]
Thanks for the explanation, you're right.

After further investigation, I see that the problem has been solved since the last release, in commit 5b7b8bcda16, by adding checks for DLIB_USE_BLAS. Looking forward to the release 19.5 :)

Thanks for the help and suggestions.

**Comment by davisking** [comment]
No problem.  That commit doesn't entirely fix the issue though, since you
could still write code that included the MKL header while at the same time
used dlib in a mode that linked to BLAS.  So it's still worth adding more
#ifndef checks for other cblas exposing headers.
